### DISCUSSION TITLE
[Feature Request] Favorite Songs

### DISCUSSION THREAD

**Discussion by fabb** [TeamRizu/OutFox#235] [03 - Feature Requests]
**Is your feature request related to a problem? Please describe.**

I have a large library and would like to find my favorite songs quickly.

**Describe the solution you'd like**

It would be cool if songs could be marked as favorite, and later quickly be found with a filter.

**Describe alternatives you've considered**

Currently it's just possible to copy favorite songs to a custom group with a file manager, but that's impractical when playing the game with dance pads in front of the TV.

**Comment by solsang** [comment]
I need this every single time i dance!
*************************************
Made a favorite folder with name "Allfavorites" and an empty song called "Allsilient" to avoid music when i start stepmania!

I find myself moving songs back and forth to this folder or wasting time around the folders to find the ones i currently play

**Comment by ListenerJubatus** [comment]
This is something we are working on for a future release

**Comment by locopati** [comment]
Any update on this highly desired feature?

**Reply by JoseVarelaP** [reply]
This feature was already added in OutFox alpha V Pre-Release Candidate 040, back in March 7th.
With a profile, you can use the `Select + Up` combination on the music wheel to favorite a song.

There are also Lua functions for those who want to use the same system on custom wheels.
```lua
PROFILEMAN:GetProfile(player):AddSongToFavorites(song)
PROFILEMAN:GetProfile(player):RemoveSongFromFavorites(song)
PROFILEMAN:GetProfile(player):SongIsFavorite(song)
PROFILEMAN:GetProfile(player):GetFavorites()
```

**Reply by rtrind** [reply]
Hello, I am using the latest version and having trouble using the quoted functions above, like SongIsFavorite. There are errors like the one below when I am trying to call the function:
```23:59:55	WARNING		Main Thread	Error playing command:...M5/Graphics/MusicWheelItem Song NormalPart/Favorites.lua:29: attempt to call a nil value (method 'SongIsFavorite')
/Appearance/Themes/Simply-DDR-SM5/Graphics/MusicWheelItem Song NormalPart/Favorites.lua:29: unknown(self = (null),params = (null),song = (null),(*temporary) = (null),(*temporary) = (null),(*temporary) = (null),(*temporary) =  (method 'SongIsFavorite'),(*temporary) = attempt to call a nil value (method 'SongIsFavorite'))
```

I am using the latest OutFox 0.5.0 Alpha Win64 pre041 version. Both the player and song variables are present, and the GetProfile function is returning a valid profile. 

Also the favorite shortcut (Select + Up) is working, and the Favorites folder is showing normally, so my only problem is the ability to identify if the song is in the favorites to display an extra icon on the music wheel.

I tried using the SoundWaves theme code to check for these functions, but none of the 4 listed above are present in the Lua code.

Any help is appreciated.

**Reply by JoseVarelaP** [reply]
The function should be functional, here's how it's performed in alpha V:
```lua
for i,player in ipairs( PlayerNumber ) do
	local prof = PROFILEMAN:GetProfile(player)
	if prof:SongIsFavorite( pSong ) then
		c.WheelPointer:SetTarget( WheelItemsActorFrame:GetChild( "ItemBGFavorite" ) )
	end
end
```

Looking at the log, it seems that some variables in there are marked as nil. As the error says: "attempt to call a nil value (method 'SongIsFavorite'))".
It seems that your PROFILEMAN:GetProfile() is somehow returning nil instead of an actual `Profile` object or the params you have is nil. You should check for that.

Also if you could show the block of code you've made for this would be appreciated to figure out what is going on.

**Reply by rtrind** [reply]
Hmmm, you are right. It seems the GetProfile is not working. I replaced my custom code with your snippet, but the behavior persists.

Weirdly if I call `SM ( prof )` after the variable definition, it prints something:
![image](https://github.com/TeamRizu/OutFox/assets/413433/093cdc74-88b2-4188-95cb-421894f3e768)

Log line:
```
/Appearance/Themes/Simply-DDR-SM5/Graphics/MusicWheelItem Song NormalPart/Favorites.lua:70: unknown(self = (null),params = (null),song = (null),(for generator) = (null),(for state) = (null),(for control) = 1,i = 1,pl = PlayerNumber_P1,prof = (null),(*temporary) = (null),(*temporary) = (null),(*temporary) = (null),(*temporary) =  (method 'SongIsFavorite'),(*temporary) = attempt to call a nil value (method 'SongIsFavorite'))
```

This is what led me to believe the GetProfile method was working. I even tried deleting my Profile (since I've been carrying it through several Outfox versions), but no luck. Same behavior.

Any ideas on why the GetProfile is not working? I am changing some files inside `Graphics\MusicWheelItem Song NormalPart`, not sure if this is relevant.

Thanks for the help so far.

**Reply by JoseVarelaP** [reply]
It should work...

Can you try just running this command on a loaded profile as is? It should give you a boolean value, or a `nil` value if the song isn't installed.

```lua
PROFILEMAN:GetProfile(PLAYER_1):SongIsFavorite( SONGMAN:FindSong('A song here by it\'s path') )
```

![Captura de pantalla 2023-08-07 a la(s) 20 43 07](https://github.com/TeamRizu/OutFox/assets/23246027/1a8238dd-0d49-459f-9880-5fca748260b9)


A worrying thing is that from the Lua stack on the log, almost everything (including the actor itself) is reporting null?
Are you checking if the current item from favorite is an actual song? Because the wheel can report invalid items like groups.
```
self = (null),params = (null), -- The object itself is null?
song = (null),(for generator) = (null), -- your Song pointer is null?
i = 1, -- The index from the iterator (i pressume this a loop?)
pl = PlayerNumber_P1, -- The player variable
prof = (null),(*temporary) -- The profile object variable
```

**Reply by rtrind** [reply]
Tried to isolate in the most clean state I could (basically installing a clean game, with new profiles and no Simply Love theme), still no dice.

Step by step:
- Clean install of [Outfox 0.5.0 Alpha](https://github.com/TeamRizu/OutFox/releases/download/OF5.0.0-041/OutFox-AlphaV-0.5.0.0-pre041-Win64-date-20230501.exe), 64bit version, in a clean folder. All options selected during install.
- Edit the default theme `Graphics/MusicWheelItem Song NormalPart.lua` file and add your suggested line at line 1
```
SCREENMAN:SystemMessage ( PROFILEMAN:GetProfile(PLAYER_1):SongIsFavorite( SONGMAN:FindSong('Outfox/Credits') ) )
```
- Run SM, open options and create some sample profiles (Teste1, Teste2, Teste3, set Teste1 as P1 and Teste2 as P2)
- Return to main menu, start a game, select profile Teste1 and reach the select music screen. See no debug message on screen
- Get the same error on the logs: 
```22:36:16	WARNING		Main Thread	Lua runtime error: ...emes/default/Graphics/MusicWheelItem Song NormalPart.lua:1: attempt to call a nil value (method 'SongIsFavorite')
/Appearance/Themes/default/Graphics/MusicWheelItem Song NormalPart.lua:1: main((*temporary) = (null),(*temporary) = (null),(*temporary) = (null),(*temporary) =  (method 'SongIsFavorite'),(*temporary) = attempt to call a nil value (method 'SongIsFavorite'))
```

If I change line 1 to the code below, same error:
```
local value=PROFILEMAN:GetProfile(PLAYER_1):SongIsFavorite( SONGMAN:FindSong('Outfox/Credits') )
if value then
	SCREENMAN:SystemMessage ( tostring(value) )
end
```

If I try the code below, I see `Profile <numeric sequence>` on the screen:
```
local value=PROFILEMAN:GetProfile(PLAYER_1)
if value then
	SCREENMAN:SystemMessage ( tostring(value) )
end
```

If I try the code below, I see `Song <numeric sequence>` on the screen:
```
local value=SONGMAN:FindSong('Outfox/Credits')
if value then
	SCREENMAN:SystemMessage ( tostring(value) )
end
```

The only explanation I can think of is I am using a wrong version of the Outfox windows executable, which does not have the new Favorite methods. Can you please check if I forgot something in the process of isolating the error?

~~PS.: I never saw this internal console your are using inside the game to query things in realtime. If this is a special theme or accesible by a shortcut on the default theme, I would love to access it. It could really speed things up.~~ 
Nevermind, just found it. Will check to see if I manage to see anything new using the console.

Thanks again for all the help.

**Reply by rtrind** [reply]
Using console, same thing:
![image](https://github.com/TeamRizu/OutFox/assets/413433/896096e2-a0fd-49d8-8fc5-5f5152e034e6)

PS.: Not sure if PLAYER_1 is a static string or If I should replace by Teste1, in my case, but the GetProfile with Teste1 value crashes the game, so I suppose not.

**Reply by JoseVarelaP** [reply]
Hmm, this might be a Lua stack issue. It's worrying that the function is not even existent when it should be. I'll talk about this with the team.

**Comment by Scraticus** [comment]
This feature will be in 042, but use the https://github.com/TeamRizu/OutFox/issues/699 to continue troubleshooting.
